cdap-watchdog/src/main/java/io/cdap/cdap/metrics/process/MessagingMetricsProcessorManagerService.java

```diff
@@ -64,33 +78,53 @@
                                           SchemaGenerator schemaGenerator,
                                           DatumReaderFactory readerFactory,
                                           MetricStore metricStore,
+                                          MetricsWriterProvider metricsWriterProvider,
                                           Set<Integer> topicNumbers,
                                           MetricsContext metricsContext,
                                           long metricsProcessIntervalMillis,
                                           int instanceId) {
     this.metricsWriters = new ArrayList<>();
     this.metricsProcessorServices = new ArrayList<>();
+    this.cConf = cConf;
+    this.metricDatasetFactory = metricDatasetFactory;
+    this.messagingService = messagingService;
+    this.schemaGenerator = schemaGenerator;
+    this.readerFactory = readerFactory;
+    this.metricStore = metricStore;
+    this.metricsWriterProvider = metricsWriterProvider;
+    this.topicNumbers = topicNumbers;
+    this.metricsContext = metricsContext;
+    this.metricsProcessIntervalMillis = metricsProcessIntervalMillis;
+    this.instanceId = instanceId;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
     MetricStoreMetricsWriter metricsWriter = new MetricStoreMetricsWriter(metricStore);
     metricsWriter.initialize(metricsContext);
     this.metricsWriters.add(metricsWriter);
 
+    for (Map.Entry<String, MetricsWriter> metricsWriterEntry : metricsWriterProvider.loadMetricsWriters().entrySet()) {
+      MetricsWriter writer = metricsWriterEntry.getValue();
+      this.metricsWriters.add(writer);
+      writer.initialize(metricsContext);
```

Comment:
The MetricsWriter also implements `Closeable`. So you have to close them on the `shutDown` method.

Reply:
done